File: zig/medium/minimum-number-of-seconds-to-make-mountain-height-zero.zig
const std = @import("std");

fn maxHeightInTime(time: u64, worker_times: []const u64) u64 {
    var total_height: u64 = 0;

    for (worker_times) |worker_time| {
        // limit = floor(8*time/worker_time) + 1
        const limit: u64 = (8 * time) / worker_time + 1;

        const sqrt_val = std.math.sqrt(limit);
        const k: u64 = (sqrt_val - 1) / 2;

        total_height += k;
    }

    return total_height;
}

pub fn minNumberOfSeconds(mountainHeight: u64, workerTimes: []const u64) u64 {

    // fast path
    if (workerTimes.len == 1) {
        const w = workerTimes[0];
        return w * mountainHeight * (mountainHeight + 1) / 2;
    }

    var left: u64 = 1;
    var right: u64 = (1_000_000_000_000 * mountainHeight) / workerTimes.len;

    while (left < right) {
        const mid = (left + right) / 2;

        if (maxHeightInTime(mid, workerTimes) >= mountainHeight) {
            right = mid;
        } else {
            left = mid + 1;
        }
    }

    return left;
}

File: cpp/medium/minimum-number-of-seconds-to-make-mountain-height-zero.cpp
#include <vector>
#include <cmath>
using namespace std;

class Solution {
private:
    long long maxHeightInTime(long long time, const vector<int>& workerTimes) {
        long long totalHeight = 0;

        for (long long workerTime : workerTimes) {
            long long limit = (8LL * time) / workerTime + 1;
            long long k = ((long long)sqrt(limit) - 1) / 2;
            totalHeight += k;
        }

        return totalHeight;
    }

public:
    long long minNumberOfSeconds(int mountainHeight, vector<int>& workerTimes) {

        if (workerTimes.size() == 1) {
            long long w = workerTimes[0];
            return w * (long long)mountainHeight * (mountainHeight + 1) / 2;
        }

        long long left = 1;
        long long right = (1000000000000LL * mountainHeight) / workerTimes.size();

        while (left < right) {
            long long mid = left + (right - left) / 2;

            if (maxHeightInTime(mid, workerTimes) >= mountainHeight)
                right = mid;
            else
                left = mid + 1;
        }

        return left;
    }
};